AX: AXTextMarker::partialOrderByTraversal is extremely slow on large webpages

twilco opened the pull request:
#### 9ca6a02d5b6ab91a124a8b4c11b837b42b3250a3
<pre>
AX: AXTextMarker::partialOrderByTraversal is extremely slow on large webpages
<a href="https://bugs.webkit.org/show_bug.cgi?id=293257">https://bugs.webkit.org/show_bug.cgi?id=293257</a>
<a href="https://rdar.apple.com/151648013">rdar://151648013</a>

Reviewed by Joshua Hoffman.

When holding VO-Right on html.spec.whatwg.org for 120 seconds, 48.6 seconds are spent in AXTextMarker::partialOrderByTraversal.
This commit replaces that function with a new one: AXTextMarker::partialOrderByAncestry. This function walks up and keeps
track of the ancestry for the |this| and |other| text markers. When we find a shared ancestor, we get the next-most
recent ancestor (which will be different for each), and determine what child-index it is in the shared ancestor. This
gets us the same information, much faster.

The old method took 48.6 seconds to process 1800 order-comparisons on html.spec.whatwg.org. The new one takes 4.4ms.

This is also faster in the case of simpler, more common webpages:

YouTube: 0.45ms vs 18.2ms for 237 comparisons
Wikipedia: 3.8ms vs. 18.7ms for 270 comparisons

Behavior covered by existing tests, and new test. I also did manual testing on html.spec.whatwg.org, YouTube, and Wikipedia where
I RELEASE_ASSERT&apos;d that the new function computed the same thing as the old one, and did not find any divergent cases.

* LayoutTests/accessibility/mac/text-marker-ordering-expected.txt: Added.
* LayoutTests/accessibility/mac/text-marker-ordering.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::partialOrder):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::operator&lt;=&gt;):
(WebCore::AXTextMarker::partialOrderByTraversal const): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::findLast const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::textMarkerRange const):

Canonical link: <a href="https://commits.webkit.org/295312@main">https://commits.webkit.org/295312@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f0a0c6302e1f5e14bc2b7f8176c70b5f8fd5848

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24258 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14676 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109757 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55219 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106588 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24643 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79379 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107554 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94357 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59704 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12431 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54591 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12484 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112145 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31709 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88470 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32073 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88088 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22480 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32986 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10759 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27026 "Hash 7f0a0c63 for PR 45617 does not build (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31636 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36976 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31428 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34767 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32988 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1579,9 +1579,102 @@ bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
 }
 
-std::partial_ordering AXTextMarker::partialOrderByTraversal(const AXTextMarker& other) const
+// This function implements a fast way to determine our ordering relative to |other| is to find
```
Nit: sentence is unclear, how about:

// This function implements a fast way to determine our ordering relative to |other|: find...

twilco replied:
Whoops, yes this got garbled after I rewrote it once! Taking your suggestion in the latest commit.

minorninth commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1579,9 +1579,102 @@ bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
+
+    // ListHashSet chosen intentionally because it has O(1) lookup time. This is important
+    // because we need to repeatedly query these lists, once every time we find a new ancestor.
+    ListHashSet<Ref<AXCoreObject>> ourAncestors;
```
Taking refs is expensive, would it work to make a hash set of AXIDs?

twilco replied:
I tried to make this change, but I actually think the existing implementation may be better for a few reasons:
* If we only stored AXIDs, we'd have to re-hydrate the AXIDs into objects later when we wanted to convert them into `index-in-parent`s. This would require also re-hydrating an `AXIsolatedTree` or `AXObjectCache` from a `AXID treeID`, which requires taking a lock (`AXTreeStore::s_storeLock`).
* It introduces a potential source of bugs — what if, somehow, the AXID we stored can no longer be turned back into an object? It probably shouldn't be possible for this to happen (especially for the isolated tree)...but it becomes possible, where it isn't with the current approach.
* This makes the code a bit less concise, adding more lines to re-hydrate several objects, and null-checks in case the situation from the previous bullet happens (or risk an unchecked deref, not ideal either)

I think with https://github.com/WebKit/WebKit/pull/45531 having landed, the refs and derefs here aren't too bad. Not set in stone on this though if you feel strongly about trying this other approach out!

minorninth commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1579,9 +1579,102 @@ bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
+// smaller-accessibility-tree pages:
+//   - YouTube: 0.45ms vs 18.2ms in 237 comparisons
+//   - Wikipedia: 3.8ms vs. 18.7ms in 270 comparisons
+std::partial_ordering AXTextMarker::partialOrderByAncestry(const AXTextMarker& other) const
```
Do we need partial ordering anywhere else in the code? I wonder if we could have partial ordering by AXCoreObject or AXIsolatedObject, and then a partial ordering of text markers would be a trivial wrapper around that.

Do you want to handle the case where the two text markers share a node but have different text indexes.

twilco replied:
> Do we need partial ordering anywhere else in the code? I wonder if we could have partial ordering by AXCoreObject or AXIsolatedObject, and then a partial ordering of text markers would be a trivial wrapper around that.

There are other places I could see this being used (the one I'm thinking of is actually in a different AXTextMarker function though). I think it's probably more logical for it to be a method on AXCoreObject though — let me try this out.

> Do you want to handle the case where the two text markers share a node but have different text indexes.

The implicit expectation is that this case is already handled by the time this function is called, specifically by `AXTextMarker::partialOrder(const AXTextMarker& other)` which is public and calls this new private function. But I think it's a good idea to make this assumption explicit with an ASSERT — will add.

minorninth reviewed (commented):
This is awesome!

I really wish WebKit had unit tests. Could you maybe try to exercise this directly from a layout test so that we can be 100% sure it works?

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1579,9 +1579,102 @@ bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
+//
+// Imagine we want to determine the ordering of D relative to G. They share ancestor A. D's descends from B
+// who is index 0 in shared ancestor A. G descends from C, who is index 1 in the shared ancestor. This lets
+// us know B (who has ancestor index 0) comes before G (who has ancestor index 1).
```
Appreciate this comment/diagram!

hoffmanjoshua commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -1579,9 +1579,102 @@ bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
+    RefPtr otherCurrent = other.object();
+    RefPtr originalOther = otherCurrent;
+    // This function should not have been called with either marker being invalid.
+    ASSERT(current && otherCurrent);
```
Should we assert the text markers before creating all of these RefPtrs? E.g.,
```cpp
ASSERT(isValid() && other.isValid())
```

twilco replied:
I like the current ASSERT as it's a bit more specific — the algorithm following cares specifically about having objects, not about the state of the text markers. But open to other ideas if you feel strongly about this one :)

twilco replied:
Though I realize now my code comment is what caused this confusion 😅 

> // This function should not have been called with either marker being invalid.

I'll just change it.

hoffmanjoshua commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1543,6 +1543,101 @@ AXCoreObject* AXCoreObject::parentObjectUnignored() const
+        }
+    }
+
+    ASSERT_NOT_REACHED();
```
Nit: could we use the text marker assert here?
```cpp
TEXT_MARKER_ASSERT_NOT_REACHED("partialOrder");
```

twilco replied:
This isn't a text marker method anymore, so it may not make sense to use that one.

hoffmanjoshua replied:
Ah, good point! Forgo that then :)

minorninth reviewed (approved):
lgtm! Nice work

twilco committed:
AX: AXTextMarker::partialOrderByTraversal is extremely slow on large webpages
https://bugs.webkit.org/show_bug.cgi?id=293257
rdar://151648013

Reviewed by Joshua Hoffman.

When holding VO-Right on html.spec.whatwg.org for 120 seconds, 48.6 seconds are spent in AXTextMarker::partialOrderByTraversal.
This commit replaces that function with a new one: AXTextMarker::partialOrderByAncestry. This function walks up and keeps
track of the ancestry for the |this| and |other| text markers. When we find a shared ancestor, we get the next-most
recent ancestor (which will be different for each), and determine what child-index it is in the shared ancestor. This
gets us the same information, much faster.

The old method took 48.6 seconds to process 1800 order-comparisons on html.spec.whatwg.org. The new one takes 4.4ms.

This is also faster in the case of simpler, more common webpages:

YouTube: 0.45ms vs 18.2ms for 237 comparisons
Wikipedia: 3.8ms vs. 18.7ms for 270 comparisons

Behavior covered by existing tests, and new test. I also did manual testing on html.spec.whatwg.org, YouTube, and Wikipedia where
I RELEASE_ASSERT'd that the new function computed the same thing as the old one, and did not find any divergent cases.

* LayoutTests/accessibility/mac/text-marker-ordering-expected.txt: Added.
* LayoutTests/accessibility/mac/text-marker-ordering.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::partialOrder):
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::operator<=>):
(WebCore::AXTextMarker::partialOrderByTraversal const): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::findLast const):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::textMarkerRange const):

Canonical link: https://commits.webkit.org/295312@main
